### PR TITLE
[Fixes #285] Allow calling pydoctor multiple times in Sphinx extension.

### LINKED ISSUE #285: Output samples
@adiroiban [wrote](https://github.com/twisted/pydoctor/pull/282#pullrequestreview-537405656):
> My suggestion is to create a pydoctor.test.sample.epydoc or pydoctor.samples.epydoc package in which we include all the possible usage of pydoctor with epydoc
>
> In this way, the result of this PR could be observed via
> https://pydoctor--282.org.readthedocs.build/en/282/api/pydoctor.test.html

@mthuurne [wrote](https://github.com/twisted/pydoctor/pull/282#issuecomment-732943736):
> I think it would be good to have a package that displays all the supported use cases, but I think it should be part of the documentation rather than part of pydoctor's code.
>
> One reason is that we'd want to showcase syntax that is not available in all Python versions, such as position-only parameters. Having that syntax in the main source dir could cause problems with distros that pre-optimize Python modules on install.
>
> Another reason is that I wouldn't want samples to be mixed into the indexes of the actual API docs. In the package index it would be just one line, but in the class hierarchy and name index it would produce more noise.

### PR DESCRIPTION
**Pull request by adiroiban**
This tries to fix #285 

I expect that 99% of pydoctor users will want to call pydoctor a single time.
This is why for those 99% cases, the interface is just a simple list.

for the other 1% you can configure the extension with a dict.

Please suggest better names for package / directories.

For now, it only provides a limited sample for epydoc/epytext.
We can extend the sample as we work on various PRs.

Support for RST can be added  in a separate PR. I am not interested in working on RST :)

Result at https://pydoctor--314.org.readthedocs.build/en/314/
https://pydoctor--314.org.readthedocs.build/en/314/docformat/epytext/demo/index.html

### CODE REVIEW

**Comment by tristanlatr** [comment]
Little remark: I would not put the links to the demo in the quick links TOC, I think it's not that primordial. We could have the links in the respective docformat pages perhaps? At the beginning.

**Comment by adiroiban** [comment]
I have reduced the scope of this PR ....

We now only have https://pydoctor--314.org.readthedocs.build/en/314/docformat/epytext/demo/index.html

This is ready for review.

**Review by mthuurne — APPROVED** [review]
It would be useful to have source links in the generated docs. I'm not sure if we can easily do relative source links though and predicting absolute URLs might complicate the build. I guess in the worst case you could pass a dummy base URL though and then replace it in the output with `sed`.

Regarding the `pydoctor_args` config item: in general I'm not a big fan of values that can have more than one type. But in this case, it is indeed likely that the vast majority of users will only run `pydoctor` once and requiring the more complex form from them would not be user-friendly. So I think having two types is fine here.

I'm approving this because I think it doesn't need a second review round. The `dict` to `Mapping` change should be done before merging though, in my opinion. The formatting and spelling fixes would be nice to include but could also be done later as they don't impact the interface.

File: docs/epytext_demo/__init__.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    - Text is shown in a monospaced font.
+    - Inline markup is not detected.
+
+Literal blocks are introduced by paragraphs ending in the special sequence "::".
+Literal blocks end at the first line whose indentation is equal to or less than that of the paragraph that introduces them.
+
+The following is a literal block::
+
+    Literal /
+           / X{Block}
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:86]: This is on purpose... as a test that we really have literal blocks... and that the X tag is not interpreted.

[the remark this answers is not in the document]
Comment (mthuurne) [reply docs/epytext_demo/__init__.py:86]: No, it's fine in that case. Reviewing epytext that is also documenting epytext is just confusing ;)

File: docs/epytext_demo/__init__.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    - Make  it easy to review HTML rendering.
+
+    - Cover all Epytext markup.
+    Like the usage of list with various indentation types.
+
+    - Have it build as part of our continuous integration tests.
+      To ensure we don't introduce regressions.
+
+
+General Epytext formating markups are documented in the docstring of this packge.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:19]: I have updated... hope I fixed all this type of errors.

File: docs/epytext_demo/__init__.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+
+Doctest Blocks
+==============
+
+    - contain examples consisting of Python expressions and their output.
+    - can be used by the doctest module to test the documented object.
+    - begin with the special sequence ">>>".
+    - are delimited from surrounding blocks by blank lines.
+    - may not contain blank lines.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/__init__.py:96]: hope this is fixed :) thanks

File: docs/epytext_demo/__init__.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+Lists
+=====
+
+Epytext supports both ordered and unordered lists.
+A list consists of one or more consecutive list items with the same indentation.
+Each list item is marked by a bullet.
+The bullet for unordered list items is a single dash character (-).
+Bullets for ordered list items consist of a series of numbers followed by periods,
+such as 12. or 1.2.8..
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/__init__.py]: I don't know :)

File: docs/epytext_demo/demo_epytext_module.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    @param m: The slope of the line.
+    @type  b: number
+    @param b: The y intercept of the line..
+    @rtype:   number
+    @return:  the x intercept of the line M{y=m*x+b}.
+    """
+    return -b/m
+
+
+def demo_typing_arguments(m: str, b: bytes) -> bool:
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:24]: I tried to update the examples. Feel free to make any changes here.

File: docs/epytext_demo/demo_epytext_module.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    @param b: The y intercept of the line.
+    @return:  the x intercept of the line M{y=m*x+b}.
+    """
+    return bool(True)
+
+
+def demo_cross_reference() -> None:
+    """
+
+    The inline markup construct LE{lb}text<object>E{rb} is used to create links to the documentation for other Python objects.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:38]: I will give it a try.

File: docs/epytext_demo/demo_epytext_module.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    'text' is the text that should be displayed for the link, and 'object' is the name of the Python object that should be linked to.
+
+    If you wish to use the name of the Python object as the text for the link, you can simply write L{object}``.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply docs/epytext_demo/demo_epytext_module.py:41]: Yep. Thanks

File: pydoctor/sphinx_ext/build_apidocs.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 Inside the Sphinx conf.py file you need to define the following configuration options:
 
   - C{pydoctor_args} - a list with all the pydoctor command line arguments used to trigger the build.
+                     - a dict with values as list of pydoctor command line arguments.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply pydoctor/sphinx_ext/build_apidocs.py:10]: I would like to keep  the dict usage ... as semi-private. Most project should not need it.
Maybe not even documenting this at all here.

File: docs/epytext_demo/__init__.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    - Make  it easy to review HTML rendering.
+
+    - Cover all epytext markup.
+    Like the usage of list with various indentation types.
+
+    - Have it build as part of our continuous integration tests.
+      To ensure we don't introduce regressions.
+
+
+General epytext formating markups are documented in the docstring of this packge.
```
Comment (mthuurne) [inline docs/epytext_demo/__init__.py:20]:
```suggestion
General epytext formating markups are documented in the docstring of this package.
```

File: docs/epytext_demo/__init__.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+Epydoc code related formating are demonstrating in the modules of this package.
```
Comment (mthuurne) [inline docs/epytext_demo/__init__.py:22]:
```suggestion
Epydoc code related formating are demonstrated in the modules of this package.
```

File: pydoctor/sphinx_ext/build_apidocs.py [hunk 11]
```diff
@@ -42,11 +43,29 @@ def on_build_finished(app: Sphinx, exception: Exception) -> None:
...
+
+    if not isinstance(runs, Mapping):
+        # We have a single pydoctor call
+        runs = {'main': runs}
+
+    for key, value in runs.items():
+        _run_pydoctor(key, value, placeholders.copy())
+
+
+def _run_pydoctor(name: str, arguments: List[str], placeholders: Mapping[str, str]) -> None:
```
Comment (mthuurne) [inline pydoctor/sphinx_ext/build_apidocs.py:56]: Do you need to modify the arguments? If not, it's better to use an immutable type such as `Sequence` or `Iterable`.

**Review by mthuurne — COMMENTED** [review]
I found a few remaining typos and one case where an annotation could maybe be made more abstract.

**Pull request merged**